cargo-verify: tweak verbosity

alastairreid opened the pull request:
This incomplete change causes commands to only print stdout/stderr if their level
is below the current verbosity.

To finish this off, I (Alastair) plan to add -q/--quiet support as well.

Commit:
cargo-verify: tweak verbosity

This change causes commands to only print stdout/stderr if their level
is below the current verbosity

Commit:
cargo-verify: bugfix: llvm-link output need not be utf-8

Commit:
cargo-verify: add --quiet mode

fshaked commented on cargo-verify/src/main.rs:
```diff
@@ -274,6 +279,9 @@ fn process_command_line() -> CVResult<Opt> {
     opt.jobs = opt.jobs_arg.unwrap_or(num_cpus::get());
 
+    // Normal mode is verbosity 1, quiet is 0
+    opt.verbose = if opt.quiet { 0 } else { opt.verbose + 1 };
```
This should be done earlier (before the use of `info_at!` above).

alastairreid replied:
Good catch

fshaked commented on cargo-verify/src/main.rs:
```diff
@@ -413,13 +421,13 @@ fn verify(opt: &Opt, package: &str, target: &str) -> CVResult<Status> {
 
     println!(
-        "test result: {:#}. {} passed; {} failed",
+        "\ntest result: {:#}. {} passed; {} failed",
```
The `\n` at the beginning needs to be documented. I'm guessing it's here to close the `print!` from `verifier_run` in the -q case?

alastairreid replied:
Good idea - done

fshaked commented on cargo-verify/src/main.rs:
```diff
@@ -438,7 +446,23 @@ fn verifier_run(opt: &Opt, bcfile: &Path, name: &str, entry: &str) -> Status {
+        let s = match status {
+            Status::Unknown => Colour::Yellow.paint("?"),
+            Status::Verified => Colour::Green.paint("."),
+            Status::Error => Colour::Red.paint("F"),
```
Maybe E instead of F?

alastairreid replied:
I chose F and . (and the colours) to match cargo-test.

fshaked commented on cargo-verify/src/main.rs:
```diff
@@ -438,7 +446,23 @@ fn verifier_run(opt: &Opt, bcfile: &Path, name: &str, entry: &str) -> Status {
+    } else {
+        println!("test {} ... {:#}", name, status);
+    }
+    term.flush();
```
error[E0425]: cannot find value `term` in this scope

alastairreid replied:
Fixed. (I'm not sure if this was a version issue - because I remember seeing that this fixed the problem (of no output until all tests completed)

fshaked commented:
I don't like that the colours can't be switched off. This makes greping harder and redirected files look bad. It doesn't seem like ansi-term supports switching the colours off. I find it very surprising considering how popular this crate is.

Commit:
cargo-verify: use enum for verbosities

Naming the different verbosity levels makes it easier
to understand the output level choices.

alastairreid commented:
oops, missed your comment about turning colours off.
Maybe there's another crate that supports it?

(Also, you will notice that I replaced all those magic verbosity numbers with an enum - hopefully easier to understand as we look at the code.)

fshaked commented:
There's yansi that can enable and disable colours on the fly, but it can't
check for terminal features (a bit old and not so popular?). You can use
https://docs.rs/console/0.14.1/console/ to check for features, and it also
has colour support.

On Thu, Mar 18, 2021 at 3:14 PM Alastair Reid ***@***.***>
wrote:

> oops, missed your comment about turning colours off.
> Maybe there's another crate that supports it?
>
> (Also, you will notice that I replaced all those magic verbosity numbers
> with an enum - hopefully easier to understand as we look at the code.)
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/project-oak/rust-verification-tools/pull/113#issuecomment-802009825>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACTBFFBVYDAVPVY7AGB2OULTEIKG7ANCNFSM4ZKPFQ6Q>
> .
>
